src/host/services/IMUService.cpp

```diff
@@ -26,38 +26,125 @@
 #include "common/signalk/SKUpdateStatic.h"
 #include "common/signalk/SKUnits.h"
 #include "IMUService.h"
+#include <EEPROM.h>
 
 void IMUService::setup() {
+  _offsetRoll  = 0.0;
+  _offsetPitch = 0.0;
+
+  // until config for positioning is made, declare default mounting position KBox here
+  uint8_t axisConfig = 0b00001001;
+  uint8_t signConfig = 0b00000000;
+
   DEBUG("Initing BNO055");
-  if (!bno055.begin()) {
+  if (!bno055.begin(bno055.OPERATION_MODE_NDOF, axisConfig, signConfig)) {
     DEBUG("Error initializing BNO055");
   }
   else {
-    DEBUG("Success!");
+    if (recallCalibration())
+      DEBUG("Success!");
   }
 }
 
 void IMUService::loop() {
-  bno055.getCalibration(&sysCalib, &gyroCalib, &accelCalib, &magCalib);
+  bno055.getCalibration(&_sysCalib, &_gyroCalib, &_accelCalib, &_magCalib);
+  //DEBUG("Calib Sys: %i Accel: %i Gyro: %i Mag: %i", _sysCalib, _accelCalib, _gyroCalib, _magCalib);
 
   eulerAngles = bno055.getVector(Adafruit_BNO055::VECTOR_EULER);
-
-  //DEBUG("Calib Sys: %i Accel: %i Gyro: %i Mag: %i", sysCalib, accelCalib, gyroCalib, magCalib);
-  //DEBUG("Attitude roll: %f pitch: %f  Mag heading: %f", eulerAngles.z(), eulerAngles.y(), eulerAngles.x());
+  //DEBUG("Sensor Raw-datas: eulerAngles.z=%.3f eulerAngles.y=%.3f  eulerAngles.x=%.3f", eulerAngles.z(), eulerAngles.y(), eulerAngles.x());
 
   SKUpdateStatic<2> update;
   // Note: We could calculate yaw as the difference between the Magnetic
   // Heading and the Course Over Ground Magnetic.
 
-  /* if orientation == MOUNTED_ON_PORT_HULL */
-  double roll, pitch, heading;
-  roll = eulerAngles.z();
-  pitch = eulerAngles.y();
-  heading = fmod(eulerAngles.x() + 270, 360);
+  // TODO: different calcs according to mounting position
+  /* if orientation == MOUNTED_ON_STARBOARD_HULL */
+  _roll = SKDegToRad(eulerAngles.z());
```

Comment:
why dont you apply the offset here?

Reply:
I need the original but "turned for mounting" values to calculate an offset.
When I apply the offset here it works just for the first time the button is pressed.
If the button is pressed again, then an offset from an offset would be calculated.